qiskit_machine_learning/kernels/statevector_kernel.py

```diff
@@ -0,0 +1,191 @@
+                      elements are fully evaluated, e.g., for two identical samples in the
+                      dataset. When inferring, all elements are evaluated. This is the default
+                      value.
+                    - ``none`` when training the diagonal is set to `1` and if two identical samples
+                      are found in the dataset the corresponding matrix element is set to `1`.
+                      When inferring, matrix elements for identical samples are set to `1`.
+            statevector_type: The type of Statevector that will be instantiated using the
+                ``feature_map`` quantum circuit and used to compute the fidelity kernel. This type
+                should inherit from and defaults to :class:`~qiskit.quantum_info.Statevector`.
+
+        Raises:
+            ValueError: When unsupported value is passed to `evaluate_duplicates`.
+        """
+        super().__init__(feature_map=feature_map)
+
+        eval_duplicates = evaluate_duplicates.lower()
+        if eval_duplicates not in ("all", "off_diagonal", "none"):
+            raise ValueError(
+                f"Unsupported value passed as evaluate_duplicates: {evaluate_duplicates}"
+            )
+        self._evaluate_duplicates = eval_duplicates
+
+        self._statevector_type = statevector_type
+
+        self._statevector_cache: dict[tuple[float, ...], np.ndarray] = {}
+
+    def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray | None = None) -> np.ndarray:
+
+        x_vec, y_vec = self._validate_input(x_vec, y_vec)
+
+        # determine if calculating self inner product
+        is_symmetric = True
+        if y_vec is None:
+            y_vec = x_vec
+        elif not np.array_equal(x_vec, y_vec):
+            is_symmetric = False
+
+        kernel_shape = (x_vec.shape[0], y_vec.shape[0])
+
+        x_svs = [self._get_statevector(x) for x in x_vec]
+        y_svs = [self._get_statevector(y) for y in y_vec]
+
+        kernel_matrix = np.ones(kernel_shape)
+        for i, x in enumerate(x_svs):
+            for j, y in enumerate(y_svs):
+                if self._is_trivial(i, j, x, y, is_symmetric):
+                    continue
+                kernel_matrix[i, j] = self._compute_kernel_element(x, y)
+
+        # due to truncation and rounding errors we may get complex numbers
+        kernel_matrix = np.real(kernel_matrix)
+
+        return kernel_matrix
+
+    @staticmethod
+    def _compute_kernel_element(x: Statevector, y: Statevector) -> float:
+        return np.abs(np.conj(x) @ y) ** 2
+
+    def _get_statevector(self, param_values) -> Statevector:
+        param_values = tuple(param_values)
+        statevector = self._statevector_cache.get(param_values, None)
+
+        if statevector is None:
+            qc = self._feature_map.bind_parameters(param_values)
+            statevector = self._statevector_type(qc).data
```

Comment:
Here this returns a `numpy.array` not a `Statevector` -- is the typehint wrong or do you want to cache the `Statevector` here? 😄

Reply:
The typehint is wrong. Thanks!